src/foam/nanos/ndiff/NDiffDAO.js

```diff
@@ -0,0 +1,72 @@
+/**
+ * @license
+ * Copyright 2021 The FOAM Authors. All Rights Reserved.
+ * http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+foam.CLASS({
+    package: 'foam.nanos.ndiff',
+    name: 'NDiffDAO',
+    extends: 'foam.dao.ProxyDAO',
+    javaIncludes: [
+        'foam.nanos.ndiff.NDiff',
+        'foam.dao.DAO',
+        'foam.dao.CompositeDAO',
+        'foam.core.X'
+    ],
+    javaCode: `
+        public NDiffDAO(foam.core.X x, foam.dao.DAO delegate) {
+            setX(x);
+            setDelegate(delegate);
+
+            foam.nanos.logger.Logger logger = (foam.nanos.logger.Logger) getX().get("logger");
+            logger.info("NDiffDAO: object created");
+        }
+    `,
+    methods: [
+        {
+            name: 'put_',
+            javaCode: `
+            foam.nanos.logger.Logger logger = (foam.nanos.logger.Logger) getX().get("logger");
+
+            if ( obj instanceof NDiff ) {
+                return super.put_( x, obj );
+            }
+
+            Object objectId = obj.getProperty("id");
+
+            // WRONG! can't grab nspec name at the moment!
+            // only temporary, will change soon
+            String nSpecName = obj.getClass().getName();
+            String dbg = "(nSpecName="+nSpecName+",objectId="+objectId+")";
```

Comment:
continuing on logging, don't to the concatenation. it will occur only if debug is enabled, for example. And we are good with , separating key/value pairs.
so assuming logger setup with Loggers
logger.info("nSpecName", nSpecName, "objectId", objectid, "find", ...);